src/functional/iterator/transforms/global_tmps.py

```diff
@@ -1,25 +1,159 @@
 from functional.iterator.transforms.popup_tmps import PopupTmps
+from functional.iterator.transforms.prune_closure_inputs import PruneClosureInputs
 
 
-class CreateGlobalTmps(NodeTranslator):
-    @staticmethod
-    def _extend_domain(domain: ir.FunCall, offset_provider, shifts):
-        assert isinstance(domain.fun, ir.SymRef) and (
-            domain.fun.id == "cartesian_domain" or domain.fun.id == "unstructured_domain"
-        )
-        assert all(isinstance(o, CartesianAxis) for o in offset_provider.values())
+AUTO_DOMAIN = ir.SymRef(id="_gtmp_auto_domain")
 
-        offset_limits = {k: (0, 0) for k in offset_provider.keys()}
 
+class Temporary(ir.Node):
+    id: Coerced[ir.SymbolName]  # noqa: A003
+    domain: Optional[ir.Expr] = None
+    dtype: Optional[Any] = None
+
+
+class FencilWithTemporaries(ir.Node, SymbolTableTrait):
+    fencil: ir.FencilDefinition
+    params: list[ir.Sym]
+    tmps: list[Temporary]
+
+
+def pformat_Temporary(printer: PrettyPrinter, node: Temporary, *, prec: int) -> list[str]:
+    start, end = [node.id + " = temporary("], [");"]
+    args = []
+    if node.domain is not None:
+        args.append(printer._hmerge(["domain="], printer.visit(node.domain, prec=0)))
+    if node.dtype is not None:
+        args.append(printer._hmerge(["dtype="], [str(node.dtype)]))
+    hargs = printer._hmerge(*printer._hinterleave(args, ", "))
+    vargs = printer._vmerge(*printer._hinterleave(args, ","))
+    oargs = printer._optimum(hargs, vargs)
+    h = printer._hmerge(start, oargs, end)
+    v = printer._vmerge(start, printer._indent(oargs), end)
+    return printer._optimum(h, v)
+
+
+def pformat_FencilWithTemporaries(
+    printer: PrettyPrinter, node: FencilWithTemporaries, *, prec: int
+) -> list[str]:
+    assert prec == 0
+    params = printer.visit(node.params, prec=0)
+    fencil = printer.visit(node.fencil, prec=0)
+    tmps = printer.visit(node.tmps, prec=0)
+    args = params + [[tmp.id] for tmp in node.tmps]
+
+    hparams = printer._hmerge([node.fencil.id + "("], *printer._hinterleave(params, ", "), [") {"])
+    vparams = printer._vmerge(
+        [node.fencil.id + "("], *printer._hinterleave(params, ",", indent=True), [") {"]
+    )
+    params = printer._optimum(hparams, vparams)
+
+    hargs = printer._hmerge(*printer._hinterleave(args, ", "))
+    vargs = printer._vmerge(*printer._hinterleave(args, ","))
+    args = printer._optimum(hargs, vargs)
+
+    fencil = printer._hmerge(fencil, [";"])
+
+    hcall = printer._hmerge([node.fencil.id + "("], args, [");"])
+    vcall = printer._vmerge(printer._hmerge([node.fencil.id + "("]), printer._indent(args), [");"])
+    call = printer._optimum(hcall, vcall)
+
+    body = printer._vmerge(*tmps, fencil, call)
+    return printer._vmerge(params, printer._indent(body), ["}"])
+
+
+PrettyPrinter.visit_Temporary = pformat_Temporary  # type: ignore
+PrettyPrinter.visit_FencilWithTemporaries = pformat_FencilWithTemporaries  # type: ignore
```

Comment:
We need to find a better place for these. Let's put a TODO for now and document somewhere (maybe ADR) why we do this IR extension.